virtool/jobs/data.py

```diff
@@ -101,33 +76,97 @@ async def _find_beta(self, query: MultiDictProxy) -> Document:
-                            "progress": "$last_status.progress",
-                            "state": "$last_status.state",
-                            "stage": "$last_status.stage",
-                        }
-                    },
-                    {"$match": {"state": {"$in": states}} if states else {}},
-                    {
-                        "$project": {
+        try:
+            page = int(query["page"])
+        except (KeyError, ValueError):
+            page = 1
+
+        try:
+            per_page = int(query["per_page"])
+        except (KeyError, ValueError):
+            per_page = 25
+
+        skip_count = 0
+
+        if page > 1:
+            skip_count = (page - 1) * per_page
+
+        sort = {"created_at": 1}
+
+        async for paginate_dict in self._db.jobs.aggregate(
+            [
+                {
+                    "$facet": {
+                        "total_count": [
+                            {"$count": "total_count"},
+                        ],
+                        "found_count": [
+                            {
+                                "$match": {
+                                    **(
+                                        compose_regex_query(term, ["user.id"])
+                                        if term
+                                        else {}
+                                    ),
+                                    **(
+                                        {"archived": archived}
+                                        if archived is not None
+                                        else {}
+                                    ),
+                                }
+                            },
+                            {
+                                "$set": {
+                                    "last_status": {"$last": "$status"},
+                                }
+                            },
+                            {
+                                "$set": {
+                                    "state": "$last_status.state",
+                                }
+                            },
+                            {"$match": {"state": {"$in": states}} if states else {}},
+                            {"$count": "found_count"},
+                        ],
+                        "data": [
+                            {
+                                "$match": {
+                                    **(
+                                        compose_regex_query(term, ["user.id"])
+                                        if term
+                                        else {}
+                                    ),
+                                    **(
+                                        {"archived": archived}
+                                        if archived is not None
+                                        else {}
+                                    ),
+                                }
```

Comment:
Hmm. Is there no way to avoid doing this twice?

If not, you can at least define the query prior to the aggregation pipeline and pass the query variable as the value for `$match` in each facet.

Reply:
I checked for a while but the only thing I could find one would be two separate aggregation calls. The first just for the total_count, and the second can use the match to both facets. I'm not sure if this would be better though.

Reply:
OK. Thanks for checking.